v2/metrics/datadog.go

```diff
@@ -0,0 +1,323 @@
+	ds.cancelSubmitLoop = cancel
+	go ds.submitLoop(submitLoopCtx)
+	return c
+}
+
+type DatadogClientOpt func(*DatadogClient)
+
+func WithDatadogTags(tags []string) DatadogClientOpt {
+	return func(c *DatadogClient) {
+		c.client.(*datadogStatsd).tags = tags
+	}
+}
+
+func WithDatadogURL(datadog url.URL) DatadogClientOpt {
+	return func(c *DatadogClient) {
+		c.cfg.Host = datadog.Host
+		c.cfg.Scheme = datadog.Scheme
+	}
+}
+
+func WithDatadogFrozenClock(frozenTime float64) DatadogClientOpt {
+	return func(c *DatadogClient) {
+		c.client.(*datadogStatsd).now = func() float64 { return frozenTime }
+	}
+}
+
+func WithDatadogSubmitInterval(dur time.Duration) DatadogClientOpt {
+	return func(c *DatadogClient) {
+		c.client.(*datadogStatsd).submitInterval = dur
+	}
+}
+
+func (d *DatadogClient) Close() error {
+	d.client.(*datadogStatsd).Close()
+	return nil
+}
+
+// datadogStatsd is an alternative statsd.Client that aggregates in memory, with periodic submission to Datadog API.
+type datadogStatsd struct {
+	authCtx          context.Context
+	cancelSubmitLoop context.CancelFunc
+	submitInterval   time.Duration
+
+	metrics *datadog.MetricsApiService
+	events  *datadog.EventsApiService
+	tags    []string
+	now     func() float64
+
+	// series tracks all metrics, but timingData is aggregated separately to prefilter calculations for summary metrics (e.g. p95)
+	mu         sync.Mutex
+	series     []*datadog.Series
+	timingData map[string]map[float64][]float64
+}
+
+var _ statsdClient = (*datadogStatsd)(nil)
+
+const (
+	submitTimeout         = 5 * time.Second
+	defaultSubmitInterval = 10 * time.Second
+)
+
+func newDatadogStatsd(cfg *datadog.Configuration, apiKey, appKey string) *datadogStatsd {
+	client := datadog.NewAPIClient(cfg)
+	keys := map[string]datadog.APIKey{
+		"apiKeyAuth": {Key: apiKey},
+		"appKeyAuth": {Key: appKey},
+	}
+	return &datadogStatsd{
+		authCtx:        context.WithValue(context.Background(), datadog.ContextAPIKeys, keys),
+		metrics:        client.MetricsApi,
+		events:         client.EventsApi,
+		now:            func() float64 { return float64(time.Now().Unix()) },
+		timingData:     make(map[string]map[float64][]float64),
+		submitInterval: defaultSubmitInterval,
+	}
+}
+
+const (
+	timingType = "gauge"
+	countType  = "count"
+)
+
+func (d *datadogStatsd) Incr(metric string, tags []string, _ float64) error {
+	tags = append(d.tags, tags...)
+	now := d.now()
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if existing := d.findSeries(metric, tags); existing != nil {
+		// Series exists: increment or add the per-timestamp count
+		for i, p := range existing.Points {
+			if p[0] == now {
+				existing.Points[i][1]++
+				return nil
+			}
+		}
+		existing.Points = append(existing.Points, []float64{now, 1})
+		return nil
+	}
+
+	// Not found, create
+	s := datadog.NewSeries(metric, [][]float64{{now, 1}})
+	s.SetType(countType)
+	s.SetTags(tags)
+	s.SetInterval(int64(d.submitInterval.Seconds()))
```

Comment:
This came up in discussion with @rxbchen , but this value is sus...

Let's imagine a count series, with events at `t=1, t=2, t=3, t=3, t=4, t=7`, then flush.
The code buckets values by the second, resulting in Points. `[][]float64{  {1,1}, {2,1}, {3,2}, {4,1}, {7,1} }`


Per https://docs.datadoghq.com/api/latest/metrics/#submit-metrics :

> interval -  If the type of the metric is rate or count, define the corresponding interval.


I'd interpreted that to mean the submission interval: how long it's been since we last sent metrics to datadog. We flush every `10s`, so this sets the interval to `10`.

It could alternatively be interpreted as the interval of each of `.Points[]` entries: we flush buckets representing `1s` of activity, so the interval should be `1`.


If a bug exists here, I think it would manifest as `10x` or `0.1x` of the graphed values of Datadog. `interval` is not a required field, so the safest choice might be to just drop this...

Reply:
I like to believe it's the first assumption that you made: https://docs.datadoghq.com/metrics/types/?tab=count#example. However, we should test what happens when we omit the `SetInterval` func

Reply:
```suggestion
```

Per discussion, remove `interval`